### ISSUE TITLE
PS is bound in GPU:0 by default, we can't change!

### ISSUE THREAD

**Issue by DjangoPeng** [tensorflow/tensorflow#3762]
I modified the seq2seq , mnist_replica and ptb model into distributed training model. But all of them have the same problem, that is when I start the server and `server.join()` the ps. The ps was bound in GPU:0 by default, and I can't change the setting. I follow the tutorial of distributed tensorflow [here](https://www.tensorflow.org/versions/r0.10/how_tos/distributed/index.html#putting-it-all-together-example-trainer-program). 
This is part of my translate.py :
```
def train():

  # set distributed configs
  ps_hosts = ["9.91.9.130:2222"]
  worker_hosts = ["9.91.9.130:2223", "9.91.9.130:2224"]
  #worker_hosts = ["9.91.9.130:2223"]

  cluster = tf.train.ClusterSpec({"ps":ps_hosts, "worker":worker_hosts})
  server = tf.train.Server(cluster,
                            job_name=FLAGS.job_name,
                            task_index=FLAGS.task_index)
  if FLAGS.job_name == "ps":
        server.join()
  elif FLAGS.job_name == "worker":
      # Worker server 
      is_chief = (FLAGS.task_index == 0)      
      gpu_num = FLAGS.task_index + 1
      #with tf.Graph().as_default():
      with tf.device(tf.train.replica_device_setter(cluster=cluster,
          worker_device="/job:worker/task:%d/gpu:%d" % (FLAGS.task_index, gpu_num))):
      #with tf.device("/gpu:%d" % FLAGS.task_index):
          """Train a en->fr translation model using WMT data."""
```
This is the GPU info:
```
Fri Aug 12 09:17:07 2016
+------------------------------------------------------+
| NVIDIA-SMI 352.39     Driver Version: 352.39         |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           On   | 0000:06:00.0     Off |                    0 |
| N/A   67C    P0    63W / 149W |  11099MiB / 11519MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K80           On   | 0000:07:00.0     Off |                    0 |
| N/A   45C    P0    72W / 149W |  10986MiB / 11519MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla K80           On   | 0000:85:00.0     Off |                    0 |
| N/A   76C    P0    66W / 149W |  11049MiB / 11519MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla K80           On   | 0000:86:00.0     Off |                    0 |
| N/A   57C    P0    75W / 149W |    223MiB / 11519MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0      9198    C   python                                       10940MiB |
|    0     29865    C   python                                          64MiB |
|    0     29961    C   python                                          64MiB |
|    1      9198    C   python                                          64MiB |
|    1     29865    C   python                                          64MiB |
|    1     29961    C   python                                       10827MiB |
|    2      9198    C   python                                          64MiB |
|    2     29865    C   python                                       10891MiB |
|    2     29961    C   python                                          64MiB |
|    3      9198    C   python                                          64MiB |
|    3     29865    C   python                                          64MiB |
|    3     29961    C   python                                          64MiB |
+-----------------------------------------------------------------------------+
```
As you can see, the pid 9198 is the `ps server` process, cause it occupied the GPU:0, I have to bind the `worker server` in GPU:1 and GPU:2.
Could we add a function or API to set the `ps server device` by configuration or arguments?

**Comment by girving** [comment]
What happens if you set the `ps_tasks` argument to `replica_device_setter`?

**Comment by DjangoPeng** [comment]
@girving I set a cluster in one machine with 4 GPUs, and tried the `ps_tasks=1`. Then I run 4 scripts in the machine:
```
#PS0
dl@dl130:~/PJT$ python rnn/translate/translate.py --job_name=ps --task_index=0

#Worker0
dl@dl130:~/PJT$ python rnn/translate/translate.py --job_name=worker --task_index=0

#Worker1
dl@dl130:~/PJT$ python rnn/translate/translate.py --job_name=worker --task_index=1

#Worker2
dl@dl130:~/PJT$ python rnn/translate/translate.py --job_name=worker --task_index=2
```
And I changed the code:
```
  ps_hosts = ["9.91.9.130:2222"]
  worker_hosts = ["9.91.9.130:2223", "9.91.9.130:2224", "9.91.9.130:2225"]
  #worker_hosts = ["9.91.9.130:2223"]

  cluster = tf.train.ClusterSpec({"ps":ps_hosts, "worker":worker_hosts})
  server = tf.train.Server(cluster,
                            job_name=FLAGS.job_name,
                            task_index=FLAGS.task_index)
  if FLAGS.job_name == "ps":
        server.join()
  elif FLAGS.job_name == "worker":
      # Worker server 
      is_chief = (FLAGS.task_index == 0)      
      gpu_num = FLAGS.task_index
      with tf.Graph().as_default():
        with tf.device(tf.train.replica_device_setter(cluster=cluster,
            worker_device="/job:worker/task:%d/gpu:%d" % (FLAGS.task_index, gpu_num),
            #ps_device="/job:ps/task:0/cpu:0",
            ps_tasks=1)):
```
But the chief task(worker0) get errors `CUDA_ERROR_OUT_OF_MEMORY: failed to allocate 10.54G down to 370.68M`, cause the memory is allocated to the `ps task(ps0)`
```
Creating 1 layers of 10 units.
------------------------------------------------------------------------------
------------------------------------------------------------------------------
I'm in chief task!
finish creating model
E tensorflow/stream_executor/cuda/cuda_driver.cc:965] failed to allocate 10.54G (11319703296 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY
...
E tensorflow/stream_executor/cuda/cuda_driver.cc:965] failed to allocate 370.68M (388683520 bytes) from device: CUDA_ERROR_OUT_OF_MEMORY
...
E tensorflow/stream_executor/cuda/cuda_blas.cc:361] failed to create cublas handle: CUBLAS_STATUS_NOT_INITIALIZED
...
tensorflow.python.framework.errors.InternalError: Blas SGEMM launch failed : a.shape=(1, 20), b.shape=(20, 10), m=1, n=10, k=20
```
What's more, the other worker task hanging forever with the log info:
```
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:244] PoolAllocator: After 7616 get requests, put_count=3642 evicted_count=1000 eviction_rate=0.274574 and unsatisfied allocation rate=0.666229
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:256] Raising pool_size_limit_ from 100 to 110
```
And even though I solve the `memory problem` by putting the worker task in other GPU, the `stuck problem` is still exist. I have no idea about the hanging process.

**Comment by girving** [comment]
@DjangoPeng They're called parameter servers because that's where the parameters are stored.  If you have one of them, all of the variables with be divided among the 1 parameter servers, which means all of the variables will go on the same machine.

**Comment by DjangoPeng** [comment]
@girving But I can't change the ps device, I know it's bound in /GPU:0 by default. So it may cause the `conflicting memory allocation` with worker0. And why all workers hanging even though the ps is ready, is it a bug? or something else?

**Comment by bixiongxu** [comment]
@DjangoPeng what do you mean "conflicting memory allocation"? If you want to change ps location, you could try switch the launching sequence, making ps to be the last one to launch. But I have no GPU env to test, sorry if I didn't get it.

**Comment by DjangoPeng** [comment]
@bixiongxu `conflicting memory allocation` means both ps0 and worker0 apply memory in GPU0, while I have fixed it by shifting one #GPU for workers. The device placement is as below:
```
ps0: GPU0
worker0: GPU1
worker1: GPU2
worker2: GPU3
```
But now my problem is the `stuck worker process`, .I can't start training even though the first step. The  log for each worker is the same, except the chief worker.
Chief worker has no output log, but only stuck at `session.run(output_feed, input_feed)`.
```
------------------------------------------------------------------------------
------------------------------------------------------------------------------
start running session
```
All other worker:
```
------------------------------------------------------------------------------
------------------------------------------------------------------------------
start running session
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:244] PoolAllocator: After 7615 get requests, put_count=3642 evicted_count=1000 eviction_rate=0.274574 and unsatisfied allocation rate=0.666185
I tensorflow/core/common_runtime/gpu/pool_allocator.cc:256] Raising pool_size_limit_ from 100 to 110
```

**Comment by prb12** [comment]
@DjangoPeng Hopefully your original issue is fixed by #4008.  
I'm going to close this issue - please reopen the original issue is still presetn.

> But now my problem is the stuck worker process,

Please can you report unrelated issues in a separate thread.
